### PR TITLE
LXD projects

### LINKED ISSUE #2689: Implement projects
Hi all, 
I have a genuine question : i'd like to get a bit more fine grained control on people being linked to my lxd daemon.

If i understand well, if someone has a valid authent plus admin password, he can do anything on the daemon (delete anything, create anything, change daemon config, etc..). 

I'd like to find an mechanism to put some limits on some kind of groups (ex : users can snaphsot and copy, only real admins can delete, etc...). I first thought something like a ManInTheMiddle would be an approach but i guess lxc client is protected against those attacks... 

Who could propose a viable option for implementing this ? Maybe a fork of lxd daemon, emptied, and linked to user/role system. Any proxy that could magicaly do the stuff ? 
Or should i PR a feature with some possibility to add plugins to lxd daemon ? 

Thank you.

### PR DESCRIPTION
**Pull request by freeekanayaka**
Make containers, profiles, images, operations, events and logging project-aware.

The branch touches virtually all internal APIs, typically introducing a "project" parameter.

Part of the database code is now being automatically generated by "make generate", which runs a new lxd-generate program. The generated code is reasonable (in the sense that it should be quite readable), however the code of lxd-generate will need more love down the road.

The storage APIs and code are now even less DRY, but we're planning a refactoring of that area.

Closes #2689.

### CODE REVIEW

File: lxd/db/cluster/update.go [hunk 1]
```diff
@@ -44,7 +45,15 @@ var updates = map[int]schema.Update{
 }
 
 func updateFromV10(tx *sql.Tx) error {
+	// Before doing anything save the counts of all tables, so we can later
+	// check that we don't accidentally delete or add anything.
+	counts1, err := query.CountAll(tx)
+	if err != nil {
+		return errors.Wrap(err, "failed to count rows in current tables")
```
Comment (stgraber) [inline lxd/db/cluster/update.go]: Capitalization

File: lxd/db/cluster/update.go [hunk 2]
```diff
@@ -115,24 +129,193 @@ CREATE TABLE tmp_profiles (
...
+DROP TABLE images_properties_copy;
+DROP TABLE images_source_copy;
+DROP TABLE profiles_config_copy;
+DROP TABLE profiles_devices_copy;
+DROP TABLE profiles_devices_config_copy;
 `
-	_, err := tx.Exec(stmts)
+	_, err = tx.Exec(stmts)
+	if err != nil {
+		return errors.Wrap(err, "failed to add project_id column")
```
Comment (stgraber) [inline lxd/db/cluster/update.go]: Capitalization

File: lxd/db/cluster/update.go [hunk 3]
```diff
@@ -115,24 +129,193 @@ CREATE TABLE tmp_profiles (
...
+	}
+
+	counts2, err := query.CountAll(tx)
+	if err != nil {
+		return errors.Wrap(err, "failed to count rows in updated tables")
```
Comment (stgraber) [inline lxd/db/cluster/update.go]: Capitalization

File: lxd/db/cluster/update.go [hunk 4]
```diff
@@ -115,24 +129,193 @@ CREATE TABLE tmp_profiles (
...
+
+	delete(counts2, "projects")
+
+	for table, count1 := range counts1 {
+		if table == "sqlite_sequence" {
+			continue
+		}
+		count2 := counts2[table]
+		if count1 != count2 {
+			return fmt.Errorf("row count mismatch in table '%s': %d vs %d", table, count1, count2)
```
Comment (stgraber) [inline lxd/db/cluster/update.go]: Capitalization

File: lxd/db/query/count.go [hunk 5]
```diff
@@ -36,3 +38,23 @@ func Count(tx *sql.Tx, table string, where string, args ...interface{}) (int, er
 
 	return count, nil
 }
+
+// CountAll returns a map associating each table name in the database
+// with the total count of its rows.
+func CountAll(tx *sql.Tx) (map[string]int, error) {
+	tables, err := SelectStrings(tx, "SELECT name FROM sqlite_master WHERE type = 'table'")
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to fetch table names")
```
Comment (stgraber) [inline lxd/db/query/count.go]: Capitalization

File: lxd/db/query/count.go [hunk 6]
```diff
@@ -36,3 +38,23 @@ func Count(tx *sql.Tx, table string, where string, args ...interface{}) (int, er
...
+	}
+
+	counts := map[string]int{}
+	for _, table := range tables {
+		count, err := Count(tx, table, "")
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to count rows of %s", table)
```
Comment (stgraber) [inline lxd/db/query/count.go]: Capitalization

File: lxd/db/schema/query.go [hunk 7]
```diff
@@ -49,7 +49,11 @@ SELECT version FROM schema ORDER BY version
 // database.
 func selectTablesSQL(tx *sql.Tx) ([]string, error) {
 	statement := `
-SELECT sql FROM sqlite_master WHERE type = 'table' AND name NOT LIKE 'sqlite_%' AND name != 'schema' ORDER BY name
+SELECT sql FROM sqlite_master WHERE
+  (type = 'table' OR type = 'index') AND
+  name != 'schema' AND
+  name NOT LIKE 'sqlite_%'
+ORDER BY name
```
Comment (stgraber) [inline lxd/db/schema/query.go]: Make a separate commit for backporting?

Comment (freeekanayaka) [reply lxd/db/schema/query.go]: Fixed.

File: lxd/db/query/count.go (hunk 6 above)
Comment (freeekanayaka) [reply lxd/db/query/count.go]: Fixed.

File: lxd/db/query/count.go (hunk 5 above)
Comment (freeekanayaka) [reply lxd/db/query/count.go]: Fixed.

File: lxd/db/cluster/update.go (hunk 4 above)
Comment (freeekanayaka) [reply lxd/db/cluster/update.go]: Fixed.

File: lxd/db/cluster/update.go (hunk 3 above)
Comment (freeekanayaka) [reply lxd/db/cluster/update.go]: Fixed.

File: lxd/db/cluster/update.go (hunk 1 above)
Comment (freeekanayaka) [reply lxd/db/cluster/update.go]: Fixed.

File: lxd/db/query/slices.go [hunk 8]
```diff
@@ -4,8 +4,40 @@ import (
...
+
+	uris := []string{}
+
+	for rows.Next() {
+		var name string
+		err := rows.Scan(&name)
+		if err != nil {
+			return nil, errors.Wrapf(err, "Failed to scan %s URI", resource)
+		}
+		uri := fmt.Sprintf("/%s/%s/%s", version.APIVersion, resource, name)
```
Comment (stgraber) [inline lxd/db/query/slices.go]: Won't work for snapshots, backups and storage-volumes, probably needs some hardcoded special casing for those.

**Comment by freeekanayaka** [comment]
jenkins: test this please

**Comment by stgraber** [comment]
@freeekanayaka so for the project features, I think we should just keep things simple and similar to what we do for other objects. So:
 - Introduce a Config attribute that's: map[string]string
 - Use the `features.images` and `features.profiles` keys to control the two features
 - During POST, any features key that's not set to false, will be set to true for the user
 - We should allow the `user.*` key namespace too as we do everywhere else

Makes sense?

**Comment by freeekanayaka** [comment]
@stgraber re the Config attribute, sounds good to me.

**Comment by stgraber** [comment]
@freeekanayaka any chance you can do a bit of rebasing on this branch?

Both to fix the newly introduced conflicts and to also reduce the number of commits in general.
I see a number of follow-up commits you've done which could be folded into some of the original commits, making it a bit easier to review things as you work on this feature.

This also has the nice side effect of making loading Github a bit faster as it very much struggles with large commit sets (as we noticed with the clustering branch).

**Comment by freeekanayaka** [comment]
@stgraber I did the rebase, conflicts were due to the candid support branch. I also squashed as many commits I could, I had being doing that already but I'll try to keep the list short even more aggressively. It's not always easy as I change stuff that I have to change again later and sometimes conflicts are laborious to solve.

**Comment by freeekanayaka** [comment]
jenkins: test this please

**Comment by freeekanayaka** [comment]
jenkins: test this please

**Comment by freeekanayaka** [comment]
jenkins: test this please

**Comment by freeekanayaka** [comment]
jenkins: test this please

**Comment by freeekanayaka** [comment]
jenkins: test this please

**Comment by freeekanayaka** [comment]
This branch is now ready for review.

**Comment by freeekanayaka** [comment]
jenkins: test this please

**Comment by stgraber** [comment]
jenkins: test this please

**Comment by freeekanayaka** [comment]
@stgraber I've moved ./lxd-generate/ to ./shared/generate/

**Comment by stgraber** [comment]
@freeekanayaka but it won't build anymore now :)

File: lxd/container_lxc.go [hunk 9]
```diff
@@ -5492,23 +5542,23 @@ func (c *containerLXC) templateApplyNow(trigger string) error {
...
+			return errors.Wrap(err, "Failed to render template")
 		}
 
 		// Figure out the architecture
 		arch, err := osarch.ArchitectureName(c.architecture)
 		if err != nil {
 			arch, err = osarch.ArchitectureName(c.state.OS.Architectures[0])
 			if err != nil {
-				return err
+				return errors.Wrap(err, "Failed to figure architecture")
```
Comment (brauner) [inline lxd/container_lxc.go]: Maybe `Failed to detect architecture`?

File: lxd/db/cluster/update.go (hunk 1 above)
Comment (brauner) [reply lxd/db/cluster/update.go]: Was that comment supposed to indicate that `fixed` should be `Fixed`? If so it doesn't look `Fixed` to me. :)

File: lxd/db/transaction.go [hunk 10]
```diff
@@ -25,11 +28,20 @@ func (n *NodeTx) Tx() *sql.Tx {
...
 
 // NodeID sets the the node NodeID associated with this cluster transaction.
 func (c *ClusterTx) NodeID(id int64) {
 	c.nodeID = id
 }
+
+func (c *ClusterTx) stmt(code int) *sql.Stmt {
+	stmt, ok := c.stmts[code]
+	if !ok {
+		panic(fmt.Sprintf("no prepared statement registered with code %d", code))
```
Comment (brauner) [inline lxd/db/transaction.go]: s/no/No/ :)

File: lxd/devlxd.go [hunk 11]
```diff
@@ -438,7 +439,8 @@ func findContainerForPid(pid int32, d *Daemon) (container, error) {
 			parts := strings.Split(string(cmdline), " ")
 			name := strings.TrimSuffix(parts[len(parts)-1], "\x00")
 
-			return containerLoadByName(d.State(), name)
+			// FIXME[projects]: how to figure out the project name here?
+			return containerLoadByProjectAndName(d.State(), "default", name)
```
Comment (brauner) [inline lxd/devlxd.go]: How crucial is this for a correctly working `devlxd`? :)

File: lxd/profiles.go [hunk 12]
```diff
@@ -40,39 +37,49 @@ var profileCmd = Command{
...
+	var result interface{}
+	err := d.cluster.Transaction(func(tx *db.ClusterTx) error {
+		hasProfiles, err := tx.ProjectHasProfiles(project)
+		if err != nil {
+			return errors.Wrap(err, "Check project features")
+		}
+
+		if !hasProfiles {
+			project = "default"
+		}
```
Comment (brauner) [inline lxd/profiles.go]: So if the project doesn't have profiles it is the default project?

File: lxd/storage_btrfs.go [hunk 13]
```diff
@@ -2849,7 +2855,7 @@ func (s *storageBtrfs) StorageEntitySetQuota(volumeType int, size int64, data in
 	switch volumeType {
 	case storagePoolVolumeTypeContainer:
 		c = data.(container)
-		subvol = getContainerMountPoint(s.pool.Name, c.Name())
+		subvol = getContainerMountPoint("default", s.pool.Name, c.Name())
```
Comment (brauner) [inline lxd/storage_btrfs.go]: Can you give me a quick argument how to decide whether to use `default` or `Project()`? :)

File: lxc/config/file.go [hunk 14]
```diff
@@ -77,7 +85,7 @@ func (c *Config) SaveConfig(path string) error {
 	defer f.Close()
 
 	// Write the new config
-	data, err := yaml.Marshal(c)
+	data, err := yaml.Marshal(conf)
```
[the remark this answers is not in the document]
Comment (stgraber) [reply lxc/config/file.go]: Yes, this is actually a bugfix, `conf` was already setup as a filtered copy of the internal config and was clearly meant to be the source of the on-disk config, but then we ended up ignoring that struct and writing `c` instead, resulting in internal remotes being saved to disk.

File: lxc/project.go [hunk 15]
```diff
@@ -0,0 +1,539 @@
...
+	if resource.name == "" {
+		return fmt.Errorf(i18n.G("Missing project name"))
+	}
+
+	// Create the project
+	project := api.ProjectsPost{}
+	project.Name = resource.name
+	project.Config = map[string]string{}
+	project.Config["features.images"] = "true"
+	project.Config["features.profiles"] = "true"
```
[the remark this answers is not in the document]
Comment (stgraber) [reply lxc/project.go]: Changing those can be done through `lxc project edit` but the code above is still not what we should do.

Instead we should just keep project.Config set to `nil` and have the server set all feature to `true` in that case, that's so we don't need the client to know what features may or may not be available on a given server.

[the remark this answers is not in the document]
Comment (stgraber) [reply lxc/project.go]: Actually, since we've turned the features into config options, we also do need to add the usual `get`, `set` and `unset` sub-commands to `lxc project` and we should also allow the `-c` option to `lxc project create` for good measure so that the user can set those at creation time.

@freeekanayaka

File: lxc/project.go [hunk 16]
```diff
@@ -0,0 +1,539 @@
...
+	if err != nil {
+		return err
+	}
+
+	data, err := yaml.Marshal(&project)
+	if err != nil {
+		return err
+	}
+
+	fmt.Printf("%s", data)
```
[the remark this answers is not in the document]
Comment (stgraber) [reply lxc/project.go]: We don't print one in all the other commands, so I don't think so.

File: client/lxd_profiles.go [hunk 17]
```diff
@@ -24,7 +24,7 @@ func (r *ProtocolLXD) GetProfileNames() ([]string, error) {
 	names := []string{}
 	for _, url := range urls {
 		fields := strings.Split(url, "/profiles/")
-		names = append(names, fields[len(fields)-1])
+		names = append(names, strings.Split(fields[len(fields)-1], "?")[0])
```
[the remark this answers is not in the document]
Comment (stgraber) [reply client/lxd_profiles.go]: No, split will always get you one entry in your slice (the original string if no match).

File: doc/rest-api.md [hunk 18]
```diff
@@ -2240,6 +2242,124 @@ HTTP code for this should be 202 (Accepted).
 
 Attempting to delete the `default` profile will return the 403 (Forbidden) HTTP code.
 
+## `/1.0/projects`
+### GET
+ * Description: List of configuration projects
```
[the remark this answers is not in the document]
Comment (stgraber) [reply doc/rest-api.md]: Oh, that's bad sed on the profile entries, should indeed be `List of projects` for consistency

File: doc/rest-api.md [hunk 19]
```diff
@@ -2240,6 +2242,124 @@ HTTP code for this should be 202 (Accepted).
...
+Input:
+
+    {
+        "features.images": True,
+        "features.profiles": True,
+        "description": "Some description string"
+    }
+
+Same dict as used for initial creation and coming from GET. The name
+property can't be changed (see POST for that).
```
[the remark this answers is not in the document]
Comment (stgraber) [reply doc/rest-api.md]: Likely copy/paste from profiles, it means that you can use the dict you get from a `GET` call or the one you use in the initial `POST` which is virtually identical (minus the name property which is immutable in this context).

File: lxd/api_internal.go [hunk 20]
```diff
@@ -781,7 +785,7 @@ func internalImport(d *Daemon, r *http.Request) Response {
 
 		// Remove the storage volume db entry for the container since
 		// force was specified.
-		err := d.cluster.StoragePoolVolumeDelete(req.Name,
+		err := d.cluster.StoragePoolVolumeDelete("default", req.Name,
```
[the remark this answers is not in the document]
Comment (stgraber) [reply lxd/api_internal.go]: Looks like for now we always import into `default`, I suspect we'll want to add a `--project` option to `lxd import` to allow overriding that, but for now that's probably good enough.

File: lxd/api_project.go [hunk 21]
```diff
@@ -0,0 +1,363 @@
...
+			project, err := tx.ProjectGet(req.Name)
+			if err != nil && err != db.ErrNoSuchObject {
+				return errors.Wrapf(err, "Check if project %q exists", req.Name)
+			}
+
+			if project != nil {
+				return fmt.Errorf("A project named '%s' already exists", req.Name)
+			}
+
+			// TODO[projects]: Is allowing renaming non-empty projects fine?
```
[the remark this answers is not in the document]
Comment (stgraber) [reply lxd/api_project.go]: No, it's not and that's why I put the check in place initially.

We should be able to do it at some point, but not right now and will involve renaming a bunch of on-disk stuff and possibly forcing all containers be off during the operation.

File: lxd/container_lxc.go (hunk 9 above)
Comment (stgraber) [reply lxd/container_lxc.go]: `Failed to detect system architecture` would make sense, yeah, not that this ever really happens :)

File: lxd/db/transaction.go (hunk 10 above)
Comment (stgraber) [reply lxd/db/transaction.go]: Also, I hate panics, but I guess it's one of those cases where we don't have a choice?

File: lxd/devlxd.go (hunk 11 above)
Comment (stgraber) [reply lxd/devlxd.go]: Probably pretty crucial actually, not fixing this would actually be a security issue as it'd allow for a container in a project that has `security.devlxd=false` to access the /dev/lxd API of a similarly named container in the default profile (both exposing an attack surface to LXD to an untrusted container and allowing an information leak between two containers).

File: lxd/profiles.go (hunk 12 above)
Comment (stgraber) [reply lxd/profiles.go]: I'm assuming ProjectHasProfiles hear check whether the project has the features.profiles feature enabled.
If it doesn't, then the project gets its profiles from the default project, so the logic looks correct to me.

File: lxd/storage_btrfs.go (hunk 13 above)
Comment (stgraber) [reply lxd/storage_btrfs.go]: You should use Project() whenever you're dealing with a resource that can be part of a project:
 - A profile
 - An image
 - A container

In this case, you're dealing with a custom storage volume, which isn't a feature that's project aware and so can only belong to the default project.

**Comment by stgraber** [comment]
@freeekanayaka I've played with your a branch a bit over the past few days and overall it's in good shape and can hopefully be merged tomorrow.

@brauner found (or made me notice) a bunch of smaller things that need fixing, the comments above should provide details but rough list is:
 - CLI
   - Add `get`, `set` and `unset`
   - Add `-c` argument to `lxc project create`
   - Don't require the CLI to know all available features for a given server, instead have all features default to `true` unless specified as `false` in the user provided config.
 - devlxd needs fixing
 - Some errors messages to tweak (capitalization and other nits)
 - Nits in documentation

Hopefully won't take very long to fix, let me know if you'd like me to deal with the devlxd thing myself.

File: client/lxd_profiles.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (freeekanayaka) [reply client/lxd_profiles.go]:
```
    If s does not contain sep and sep is not empty, Split returns a slice of
    length 1 whose only element is s.
```

File: lxd/db/cluster/update.go (hunk 1 above)
Comment (freeekanayaka) [reply lxd/db/cluster/update.go]: Opps, true! Should be ```really-fixed``` now.

File: doc/rest-api.md (hunk 18 above)
[the remark this answers is not in the document]
Comment (freeekanayaka) [reply doc/rest-api.md]: Changed to ```List of projects``` as recommended by @stgraber. Note that we have some cases where we use ```List``` (capital) and some where we use ```list`` (lowercase). Should probably be addressed separately from this branch.

File: doc/rest-api.md (hunk 19 above)
[the remark this answers is not in the document]
Comment (freeekanayaka) [reply doc/rest-api.md]: So I presume no rewording is needed here?

File: lxd/db/transaction.go (hunk 10 above)
Comment (freeekanayaka) [reply lxd/db/transaction.go]: Changed from ```no``` to ```No```.

We indeed have no choice. It's basically impossible to trigger the panic in a normal release, this should be triggered only at development time if you are misusing the code generation tool.

File: lxd/devlxd.go (hunk 11 above)
Comment (freeekanayaka) [reply lxd/devlxd.go]: Oh yeah, I forgot to highlight this fact in the PR description. Good catch. Any idea how we could fix this?

File: lxd/profiles.go (hunk 12 above)
Comment (freeekanayaka) [reply lxd/profiles.go]: Correct, ```hasProfiles``` checks features.profiles.

File: lxd/storage_btrfs.go (hunk 13 above)
Comment (freeekanayaka) [reply lxd/storage_btrfs.go]: Nice explanation.

File: lxd/api_project.go (hunk 21 above)
[the remark this answers is not in the document]
Comment (freeekanayaka) [reply lxd/api_project.go]: Good catch, I forgot to mention this in the PR description too. I've added the check now.

File: lxd/container_lxc.go (hunk 9 above)
Comment (freeekanayaka) [reply lxd/container_lxc.go]: Yeah this is mainly a help for development. When you are changing (aka breaking) stuff and you get errors, it's important to know what code path triggered them.

File: lxc/project.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (freeekanayaka) [reply lxc/project.go]: Should be all done (new commands and default feature values).

**Comment by freeekanayaka** [comment]
@stgraber I think everything is sorted except for the devlxd issue. I'm not sure how to address that, since AFAIU from an OS point of view the only hint about the project a container belongs to would be its rootfs path. Is the rootfs path guaranteed to be exposed in /proc/<pid>/cmdline?

File: lxd/devlxd.go (hunk 11 above)
Comment (stgraber) [reply lxd/devlxd.go]: Doesn't the project appear in the `[lxc monitor] ...` process string? (as project_name)

Comment (brauner) [reply lxd/devlxd.go]: @stgraber, iff the rootfs path does have the project appended but yeah, it should. We can also introduce a new config setting (I thought about this for a while): `lxc.monitor.proctitle` or something that would let users set a name for the process.
